text/0000-package-executables.md

```diff
@@ -0,0 +1,177 @@
+
+- `notion install cowsay` - this installs `cowsay` as a user tool
+- `cowsay 'hi'` - this runs `cowsay` as expected
+- `notion install node 8` - this sets the user version of Node to 8.x.x
+- `cowsay 'hello'` - this still works as expected, and uses Node 10.1.2 to run it
+
+Again this is transparent to the user. When they install a tool, it continues to work using the version of Node that it was installed with.
+
+### Combination of project and user tools:
+
+The user toolchain includes the `madge` tool (which is useful for analyzing the module graph of Node projects). The `madge` tool requires at least Node 6.x to run.
+
+The user is using this tool to analyze a project (which is the point of the tool), but it’s not part of the scripts for that project. So it’s not a project tool.
+
+The project happens to be using Node 4.x (older than what `madge` requires), but it works just fine. `madge` is pinned to the version of Node it was installed with in the user toolchain, not the version of Node the user’s project is pinned to.
+
+Note also that the project scripts can’t accidentally depend on `madge` because it is a user tool and the scripts don’t see it in their PATH.
+
+
+## How to think about these features
+
+### Project tools
+
+Any package executables installed by direct `dependencies` and `devDependencies` of a Node project are available for the user to use when they are in that project. These executables will run using the Node version that is configured for the project.
+
+### User tools
+
+When a user installs a package executable globally, it’s like they are installing a static binary. It will work in the same way that it was installed, despite changing directories or changing Node versions.
+
+# Details
+[details]: #details
+
+To make this work, Notion will put shims on the PATH in `~/.notion/shims/` to intercept execution of package executables and delegate appropriately. These shims are implemented as symlinks to the `launchbin` executable, where the name of the symlink is the name of the shimmed executable. Since the shims directory will come first (or very early) in the PATH, when any shimmed executable is run, it will execute `launchbin` with the name of the executable as the first argument.
+
+Then `launchbin` uses this logic to determine how to delegate execution:
+
+- If the user is in a Node project, and the executable is a direct dependency of the current project, then that is executed (using the absolute path to the executable). We determine that an executable is a direct dependency by reading the `"bin"` section of each direct dependency’s `package.json`. Only direct dependencies are executed because `npm` and `yarn` install executables for the complete tree of project dependencies, not just direct dependencies. Exposing all of those would be a security risk as well as confusing to debug.
+- Otherwise this is installed as a user tool, so that is executed (using the absolute path to the executable).
```

Comment:
This case should _check_ if it's installed as a user tool. Then there's a subsequent case, which is to delegate to whatever system tool might be there in the remainder of `PATH`.